File: apps/python/face_id.py
import pathlib
import os

from oneML import faceAPI as api

os.environ['ONEML_CPP_MIN_LOG_LEVEL'] = "ERROR"
assets_path = os.path.join(
    str(pathlib.Path(__file__).resolve().parent.parent.parent.absolute()),
    "assets/images",
)


def main():
    manager = api.LicenseManager()
    manager.activate_trial()
    embedder = api.FaceEmbedder(manager)
    face_id = api.FaceId(embedder, manager)
    utils = api.Utils(manager)

    # Images
    img1 = utils.read_image_cv(
        os.path.join(
            assets_path, "register-set/Colin_Powell/colin_powell_0074.jpg"
        )
    )

    img2 = utils.read_image_cv(
        os.path.join(
            assets_path, "register-set/Colin_Powell/colin_powell_0078.jpg"
        )
    )

    img3 = utils.read_image_cv(
        os.path.join(
            assets_path, "register-set/George_Robertson/george_robertson_0000.jpg"
        )
    )

    img4 = utils.read_image_cv(
        os.path.join(
            assets_path, "register-set/George_Robertson/george_robertson_0002.jpg"
        )
    )

    # is_the_same_person_img
    is_same = face_id.is_the_same_person_img(img1, img2)
    print("Is the same person: " + str(int(is_same)))

    # is_the_same_person_emb
    embeddings = embedder.embed(img1)
    is_same, new_emb = face_id.is_the_same_person_emb(embeddings.get_embedding(), img2)
    print("Is the same person: " + str(int(is_same)))

    # register_id_emb
    size = face_id.register_id_emb("Colin_Powell", embeddings.get_embedding())
    print("Registered size is: " + str(size))

    # register_id_images
    emb = face_id.register_id_image("George_Robertson", img3)
    print(
        "Registered emb is: ["
        + "".join("{:.6f}, ".format(k) for k in emb[0:5])[:-2]
        + "]"
    )

    # predict
    result = face_id.predict(img2)

    print("Is identifiable: " + str(int(result.is_identifiable())))
    print("Id: " + str(result.get_id()))
    print(
        "Nearest node distance: "
        + "".join("{:.6f}".format(result.get_nearest_node_distance()))
    )
    print(
        "Combined distance: " + "".join("{:.6f}".format(result.get_combined_distance()))
    )

    # update_embedding_dynamically_img
    face_id.update_embedding_dynamically_img("George_Robertson", img3, emb)
    result = face_id.predict(img4)

    print("Is identifiable: " + str(int(result.is_identifiable())))
    print("Id: " + str(result.get_id()))
    print(
        "Nearest node distance: "
        + "".join("{:.6f}".format(result.get_nearest_node_distance()))
    )
    print(
        "Combined distance: " + "".join("{:.6f}".format(result.get_combined_distance()))
    )

    # update_embedding_dynamically_emb
    face_id.update_embedding_dynamically_emb(
        "Colin_Powell", new_emb, embeddings.get_embedding()
    )
    result = face_id.predict(img1)

    print("Is identifiable: " + str(int(result.is_identifiable())))
    print("Id: " + str(result.get_id()))
    print(
        "Nearest node distance: "
        + "".join("{:.6f}".format(result.get_nearest_node_distance()))
    )
    print(
        "Combined distance: " + "".join("{:.6f}".format(result.get_combined_distance()))
    )

    # update_embedding
    face_id.update_embedding("George_Robertson", emb)
    result = face_id.predict(img3)

    print("Is identifiable: " + str(int(result.is_identifiable())))
    print("Id: " + str(result.get_id()))
    print(
        "Nearest node distance: "
        + "".join("{:.6f}".format(result.get_nearest_node_distance()))
    )
    print(
        "Combined distance: " + "".join("{:.6f}".format(result.get_combined_distance()))
    )

    # remove_id
    face_id.remove_id("George_Robertson")
    face_id.remove_id("Colin_Powell")

    # get_ids
    ids = face_id.get_ids()
    print("Gallery size: " + str(len(ids)))


if __name__ == "__main__":
    main()

File: apps/golang/face_id.go
package main

import (
	"faceapps/utility"
	"fmt"
	"oneml/face"
	"os"
	"path"
)

func main() {
	currentPath, _ := os.Getwd()
	testAssetPath := path.Join(currentPath, "../../assets/images")

	// Initialize oneML modules
	licenseManager := face.NewLicenseManager()
	licenseManager.ActivateTrial()

	faceEmbedder := face.NewFaceEmbedder(licenseManager)
	faceId := face.NewFaceId(faceEmbedder, licenseManager)
	oneMLUtils := face.NewUtils(licenseManager)
	defer face.DeleteUtils(oneMLUtils)
	defer face.DeleteFaceEmbedder(faceEmbedder)
	defer face.DeleteFaceId(faceId)

	// load Image
	filePath1 := path.Join(testAssetPath, "register-set/Colin_Powell/colin_powell_0074.jpg")
	img1 := oneMLUtils.ReadImageCV(filePath1)

	filePath2 := path.Join(testAssetPath, "register-set/Colin_Powell/colin_powell_0078.jpg")
	img2 := oneMLUtils.ReadImageCV(filePath2)

	filePath3 := path.Join(testAssetPath, "register-set/George_Robertson/george_robertson_0000.jpg")
	img3 := oneMLUtils.ReadImageCV(filePath3)

	filePath4 := path.Join(testAssetPath, "register-set/George_Robertson/george_robertson_0002.jpg")
	img4 := oneMLUtils.ReadImageCV(filePath4)

	// IsTheSamePersonImg
	isSame := faceId.IsTheSamePersonImg(img1, img2)
	fmt.Println("Is the same person:", utility.BoolToInt(isSame))

	// IsTheSamePersonEmb
	embeddings := faceEmbedder.Embed(img1)
	samePersonResult := faceId.IsTheSamePersonEmb(embeddings.GetEmbedding(), img2)
	fmt.Println("Is the same person:", utility.BoolToInt(samePersonResult.GetFirst()))

	// RegisterIdEmb
	size := faceId.RegisterIdEmb("Colin_Powell", embeddings.GetEmbedding())
	fmt.Println("Registered size is:", size)

	// RegisterIdImage
	emb := faceId.RegisterIdImage("George_Robertson", img3)
	fmt.Printf("Registered emb is: [%.6f, %.6f, %.6f, %.6f, %.6f]\n", emb.Get(0), emb.Get(1), emb.Get(2), emb.Get(3), emb.Get(4))

	// Predict
	result := faceId.Predict(img2)
	fmt.Println("Is identifiable:", utility.BoolToInt(result.IsIdentifiable()))
	fmt.Println("Id:", result.GetId())
	fmt.Printf("Nearest node distance: %.6f\n", result.GetNearestNodeDistance())
	fmt.Printf("Combined distance: %.6f\n", result.GetCombinedDistance())

	// UpdateEmbeddingDynamicallyImg
	faceId.UpdateEmbeddingDynamicallyImg("George_Robertson", img3, emb)

	result = faceId.Predict(img4)
	fmt.Println("Is identifiable:", utility.BoolToInt(result.IsIdentifiable()))
	fmt.Println("Id:", result.GetId())
	fmt.Printf("Nearest node distance: %.6f\n", result.GetNearestNodeDistance())
	fmt.Printf("Combined distance: %.6f\n", result.GetCombinedDistance())

	// UpdateEmbeddingDynamicallyEmb
	faceId.UpdateEmbeddingDynamicallyEmb("Colin_Powell", samePersonResult.GetSecond(), embeddings.GetEmbedding())

	result = faceId.Predict(img1)
	fmt.Println("Is identifiable:", utility.BoolToInt(result.IsIdentifiable()))
	fmt.Println("Id:", result.GetId())
	fmt.Printf("Nearest node distance: %.6f\n", result.GetNearestNodeDistance())
	fmt.Printf("Combined distance: %.6f\n", result.GetCombinedDistance())

	// UpdateEmbedding
	faceId.UpdateEmbedding("George_Robertson", emb)

	result = faceId.Predict(img3)
	fmt.Println("Is identifiable:", utility.BoolToInt(result.IsIdentifiable()))
	fmt.Println("Id:", result.GetId())
	fmt.Printf("Nearest node distance: %.6f\n", result.GetNearestNodeDistance())
	fmt.Printf("Combined distance: %.6f\n", result.GetCombinedDistance())

	// RemoveId
	faceId.RemoveId("Colin_Powell")
	faceId.RemoveId("George_Robertson")

	// GetIds
	ids := faceId.GetIds()
	fmt.Println("Gallery size:", ids.Size())

	report := faceId.GetUsage()
	report.ToLog()
}
